content/shmem_test_all.tex

```diff
@@ -0,0 +1,63 @@
+\apisummary{
+    Indicate whether all variables within an array of variables on the local \ac{PE} meet a specified test condition.
+}
+
+\begin{apidefinition}
+
+\begin{C11synopsis}
+int @\FuncDecl{shmem\_test\_all}@(TYPE *ivars, size_t nelems, const int *status, int cmp, TYPE cmp_value);
+\end{C11synopsis}
+where \TYPE{} is one of the point-to-point synchronization types specified by
+Table \ref{p2psynctypes}.
+
+\begin{Csynopsis}
+int @\FuncDecl{shmem\_\FuncParam{TYPENAME}\_test\_all}@(TYPE *ivars, size_t nelems, const int *status, int cmp,
+    TYPE cmp_value);
+\end{Csynopsis}
+where \TYPE{} is one of the point-to-point synchronization types and has a
+corresponding \TYPENAME{} specified by Table \ref{p2psynctypes}.
+
+\begin{apiarguments}
+
+  \apiargument{IN}{ivars}{A pointer to an array of remotely accessible data
+    objects.}
+  \apiargument{IN}{nelems}{The number of elements in the \VAR{ivars} array.}
+  \apiargument{IN}{status}{An optional mask array of length \VAR{nelems}
+    that indicates which elements in \VAR{ivars} are excluded from the test set.}
+  \apiargument{IN}{cmp}{A comparison operator from Table~\ref{p2p-consts}
+    that compares elements of \VAR{ivars} with \VAR{cmp\_value}.}
+  \apiargument{IN}{cmp\_value}{The value to be compared with the objects
+    pointed to by \VAR{ivars}.}
+
+\end{apiarguments}
+
+\apidescription{
+    The \FUNC{shmem\_test\_all} routine behaves similarly to
+    \FUNC{shmem\_wait\_until\_all}, but it does not block and returns zero if
```

Comment:
I don't love cross-API references. As someone who uses the spec as an API reference, it would be annoying if I were trying to refresh my memory on shmem_test_all to then have to jump over to shmem_wait_until_all and read that, only to have to jump back. I'd much prefer shmem_test_all's description be self-contained as much as possible, even at the cost of redundant text with shmem_wait_until_all. Any objections?

Reply:
It also makes the spec harder to read as a new user.

Reply:
I believe I have heard some objections to full self-containment in the past, but perhaps not...  Might as well try taking a poll!
https://doodle.com/poll/q6chkaqv76psrttn

I'm actually slightly in favor of removing the cross-API references as well, but I would prefer to make the changes in a separate ticket.  Would that be acceptable?  If so, I can start preparing this before we do the voting on this ticket.

Reply:
I think cross references make more sense if one thing is defined in terms of extending the behavior of another. This is the reverse scenario, in which one thing is defined in terms of something else, but not exactly all that it does, which only serves to be confusing.

Specifically, the cross reference here would make more sense if `shmem_wait_until_all` was defined in terms of `shmem_test_all`, not vice versa.

Reply:
I've always wondered why `shmem_test` was defined after `shmem_wait` in the spec...  Perhaps it's easier to absorb conceptually in that order as a user, but I'd certainly be open to switching.  As a developer, `shmem_wait` is easily implemented using `shmem_test` so why not define test first?  Or we could just not worry about it and embrace self-containment ;)

Reply:
I think it's because `shmem_wait[_until]` predates OpenSHMEM 1.0, but `shmem_test` wasn't added until OpenSHMEM 1.4.

Reply:
Wow, somehow I didn't catch that `shmem_test` was new in 1.4...  It must have been proposed relatively early on.  Thanks.